test/libsolidity/ABIJsonTest.cpp

```diff
@@ -0,0 +1,103 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/**
+ * Unit tests for the solidity compiler ABI JSON Interface output.
+ */
+
+#include <test/libsolidity/ABIJsonTest.h>
+
+#include <test/Options.h>
+
+#include <libsolidity/interface/CompilerStack.h>
+#include <libdevcore/JSON.h>
+#include <libdevcore/AnsiColorized.h>
+
+#include <boost/algorithm/string.hpp>
+
+#include <fstream>
+
+using namespace std;
+using namespace dev;
+using namespace dev::solidity;
+using namespace dev::solidity::test;
+
+ABIJsonTest::ABIJsonTest(string const& _filename)
+{
+	ifstream file(_filename);
+	if (!file)
+		BOOST_THROW_EXCEPTION(runtime_error("Cannot open test contract: \"" + _filename + "\"."));
+	file.exceptions(ios::badbit);
+
+	m_source = parseSourceAndSettings(file);
+	m_expectation = parseSimpleExpectations(file);
+}
+
+TestCase::TestResult ABIJsonTest::run(ostream& _stream, string const& _linePrefix, bool _formatted)
+{
+	CompilerStack compiler;
+
+	compiler.setSources({{"", "pragma solidity >=0.0;\n" + m_source}});
+	compiler.setEVMVersion(dev::test::Options::get().evmVersion());
+	compiler.setOptimiserSettings(dev::test::Options::get().optimize);
+	if (!compiler.parseAndAnalyze())
+		BOOST_THROW_EXCEPTION(runtime_error("Parsing contract failed"));
+
+	m_obtainedResult.clear();
+	bool first = true;
+	for (string const& contractName: compiler.contractNames())
+	{
+		if (!first)
+			m_obtainedResult += "\n\n";
+		m_obtainedResult += "    " + contractName + "\n";
+		m_obtainedResult += jsonPrettyPrint(compiler.contractABI(contractName)) + "\n";
+		first = false;
+	}
+	if (m_expectation != m_obtainedResult)
```

Comment:
Would it make sense to parse the result/expectation and compare the json ast instead?

Reply:
I guess it would make it easier to highlight the differences in output for example

Reply:
If we do that, the auto-update might result in a different text even though the comparison resulted in "equal" - is that what we want?

Reply:
Also I guess it is good to know that this is all deterministic :)

Reply:
We could do both. And only run the AST comparison once the string comparsion failed ;)